[VideoOSD] Add autoclose setting

enen92 opened the pull request:
## Description
Any modern video player (or video application) autocloses the video OSD after a small period of inactivity by the user. In KODI this seems to be skin dependent. Some skins implement the feature, others do not. Personally, I think this has a broader scope than skins (it defines behaviour not visual aspect) and should be provided as an option in the application itself.

- I am even unsure if it shouldn't be enabled by default (currently it's off). What do you think?
- Also unsure about the location of the setting. Does a new category "Video OSD" make sense? It's odd to be under "Skin"

## Motivation and context
There are too many clicks involved in the OSD, usually the user just wants to watch its own media. For instance, when downloading a subtitle or selecting a subtitle, the OSD buttons will be left on the screen and the user has to explicitly hit back to watch the content.

## How has this been tested?
Runtime tested

## What is the effect on users?
Slicker user experience if the new setting is enabled

## Screenshots (if appropriate):
![image](https://user-images.githubusercontent.com/7375276/136280572-1eb9e1e5-ddb5-4fe8-8c96-a0da62c008cc.png)


## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

jjd-uk commented:
Music OSD behaves differently from Video, so if that was touch care would be to be taken to not break current behaviour. OSD/Info can be switched to permently on/off with a toggle of the appropiate button similar to video, however in the case of the Info not being displayed then Kodi will always display the Info on a track change for 10 sec before hiding again.

Ch1llb0 commented:
> I am even unsure if it shouldn't be enabled by default (currently it's off). What do you think?

> In my view Off is the correct Default as that is what emulates current behaviour.

My skin is one of those which offer this feature currently and I'd tend to say it shouldn't be 'off' by default. It's a change, so why not go all the way and also implement a new behaviour? The OSD auto hiding is something many people have requested on our forums and it seems sensible to me to make that a default - which time period to set as default would be a separate question.

enen92 commented:
@ksooo thinking a bit more about this I don't see a way of making it consistent and include PVR at the same time.
I am tempted to create a new section in Settings/Interface dedicated to Player OSD (and create a separator called "Video" inside). However, I think the PVR behaviour when switching channels likely should stay in the PVR section of settings - not everyone has PVR enabled and the behaviour that setting defines is specific to an action that affects "objects" that only have meaning in PVR (channels). Repeating the setting in both places is also an option though (I'd try to avoid it as much as possible).

Regarding music I won't be touching/changing behaviour with this PR for now but it can fit well in a "Music" separator within this new "Player OSD" main menu section.
It might look a bit empty for now with just this setting but I can see a few more that can be implemented. For example, currently with commbreaks a kaytoast is shown when a commbreak is skipped with no option to avoid it. Some requests have been made in the forum to have an option to disable it.

Ch1llb0 commented:
It would also be nice to be able to adjust the time until the music OSD is hidden... This is currently fixed and there's not setting to change it. Such a new section would make sense IMHO, @enen92 👍🏻 Put all OSD related settings for PVR, video _and_ music there.
Wouldn't this page make more sense under Settings/Player though? As it's player related, users might look there first and wonder why it's not there...

enen92 commented:
Because it's not related nor defines the player behaviour. It's only used to define the GUI operation (so the interface)...it just happens to be on the windows that have the player control (in this case the video osd).
Maybe better to call it only OSD to avoid confusion with the Player section.

enen92 committed:
[Settings][VideoOSD] Add autoclose setting

enen92 commented:
Rebased

- Created new category "OSD" under "Settings/Interface" and a setting group dedicated to "Video" (as I'm only applying this behaviour on the video osd window). Music can be added later to a new group if someone is interested.
- Changed the setting to true as default with a 3 second default value
- Applied ksooo comments

![image](https://user-images.githubusercontent.com/7375276/146043528-a4e0febb-de54-4841-9cd9-828774c5f75f.png)

Feedback appreciated

enen92 commented:
Changed default Video OSD close time to 5 seconds.
Auto close video OSD setting is a Basic level setting. Elapsed time for autoclosing the video OSD has been moved to Advanced level.
See comment above.

I plan to merge this on Sunday night if no opposition/changes are requested.

enen92 commented:
Jenkins build this please

jjd-uk commented:
Looks good from the settings xml side of things. Can't say anything about the C++.

HitcherUK commented:
@enen92 A user has raised a good point on the forum that different skins can have other dialogs open at the same time to OSD is visible and they remain open when the OSD gets closed automatically. Maybe this should close all open dialogs as well?

https://forum.kodi.tv/showthread.php?tid=367983

enen92 commented:
As far as I can tell there is no clean way to close all other dialogs when this happens from the core, specially because you don't know when they were open. Which dialogs are we talking about? The only way is to make those dialogs somehow depend on the visibility of the video osd.
I already thought a bit about this PR and somehow I think the intention was good but it doesn't belong to the core... should in fact be handled by the skins. However there is also no clean way to implement this in the skin at the moment (my idea is to have something like skin timers that would trigger builtins on elapsed time depending on Boolean conditions).
Maybe for now we should just revert this one since the first alpha is not yet out.
